### PR TITLE
iOS screen sharing support

### PR DESCRIPTION
**Pull request by alexbumbu**

### CODE REVIEW

File: BroadcastPickerView.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+'use strict';
+
+import {
+  requireNativeComponent,
+} from 'react-native';
+
+const View = requireNativeComponent('ScreenCapturePickerView');
+
+export default View;
```
Comment (saghul) [inline BroadcastPickerView.js:9]: You can skip the extra variable and just do `export default requireNativeComponent('ScreenCapturePickerView');`

File: ios/RCTWebRTC/ScreenCapturer.m [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    }
+}
+
+- (void)didCaptureVideoFrame:(CVPixelBufferRef)pixelBuffer {
+    if (!self.startDate) {
+        self.startDate = [NSDate date];
+    }
+        
+    RTCCVPixelBuffer *rtcPixelBuffer = [[RTCCVPixelBuffer alloc] initWithPixelBuffer: pixelBuffer];
+    int64_t timeStampNs = fabs([self.startDate timeIntervalSinceNow]) * kNanosecondsPerSecond;
```
Comment (saghul) [inline ios/RCTWebRTC/ScreenCapturer.m:191]: Isn't NSDate too "heavy" for this? I tnow it doesn't directly translate, but this what the WebRTC sample does:

`int64_t timeStampNs =
      CMTimeGetSeconds(CMSampleBufferGetPresentationTimeStamp(sampleBuffer)) * NSEC_PER_SEC;`

File: ios/RCTWebRTC/SocketConnection.m [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    if (self.serverSocket < 0) {
+       NSLog(@"failure creating socket");
+       return nil;
+    }
+    
+    if (![self setupSocketBufferSize]) {
+        return  nil;
+    }
+    
+    if (![[NSFileManager defaultManager] createFileAtPath:filePath contents:nil attributes:nil]) {
```
Comment (saghul) [inline ios/RCTWebRTC/SocketConnection.m:44]: Why are you creating the file just to unlink it later, before you bind?

File: ios/RCTWebRTC/SocketConnection.m [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        } while (![NSThread currentThread].isCancelled);
+    }];
+    self.networkThread.qualityOfService = NSQualityOfServiceUserInitiated;
+}
+
+- (BOOL)setupSocketBufferSize {
+    int optval = kConnectionBufferSize;
+    socklen_t optlen = sizeof(optval);
+    
+    int status = setsockopt(self.serverSocket, SOL_SOCKET, SO_SNDBUF, &optval, optlen);
```
Comment (saghul) [inline ios/RCTWebRTC/SocketConnection.m:126]: This probably has no effect at the moment. It needs to run on the connected socket, so in what you get after calling accept(), not in the server socket itself.

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m [hunk 5]
```diff
@@ -123,21 +172,17 @@ - (RTCVideoTrack *)createVideoTrack:(NSDictionary *)constraints {
...
                                                                   position:AVCaptureDevicePositionUnspecified];
     for (AVCaptureDevice *device in videoevicesSession.devices) {
-        NSString *position = @"unknown";
+        NSString *position = @"";
         if (device.position == AVCaptureDevicePositionBack) {
             position = @"environment";
         } else if (device.position == AVCaptureDevicePositionFront) {
             position = @"front";
         }
-        NSString *label = @"Unknown video device";
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:132]: Please don't undo this, I've seen crashes because it was null.

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m [hunk 6]
```diff
@@ -123,21 +172,17 @@ - (RTCVideoTrack *)createVideoTrack:(NSDictionary *)constraints {
                                                                  mediaType:AVMediaTypeVideo
                                                                   position:AVCaptureDevicePositionUnspecified];
     for (AVCaptureDevice *device in videoevicesSession.devices) {
-        NSString *position = @"unknown";
+        NSString *position = @"";
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:175]: why this change?

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m [hunk 7]
```diff
@@ -146,14 +191,10 @@ - (RTCVideoTrack *)createVideoTrack:(NSDictionary *)constraints {
                                                                  mediaType:AVMediaTypeAudio
                                                                   position:AVCaptureDevicePositionUnspecified];
     for (AVCaptureDevice *device in audioDevicesSession.devices) {
-        NSString *label = @"Unknown audio device";
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:149]: ditto

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m [hunk 8]
```diff
@@ -198,6 +239,15 @@ - (RTCVideoTrack *)createVideoTrack:(NSDictionary *)constraints {
 {
   RTCMediaStream *stream = self.localStreams[streamID];
   if (stream) {
+    for (RTCVideoTrack *track in stream.videoTracks) {
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:242]: I think you got a weird merge, this code was refactored.

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m [hunk 9]
```diff
@@ -232,7 +282,7 @@ - (RTCVideoTrack *)createVideoTrack:(NSDictionary *)constraints {
   RTCMediaStreamTrack *track = self.localTracks[trackID];
   if (track) {
     RTCVideoTrack *videoTrack = (RTCVideoTrack *)track;
-    [videoTrack.videoCaptureController switchCamera];
+    [(VideoCaptureController *)videoTrack.captureController switchCamera];
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:285]: Check the instance type here, since it could be called on a screen sharing stream.

**Review by saghul — COMMENTED** [review]
Left come comments Alex, PTAL!

File: ios/RCTWebRTC/ScreenCapturer.m [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+    
+    RTCVideoFrame *videoFrame = [[RTCVideoFrame alloc] initWithBuffer: rtcPixelBuffer
+                                                             rotation: RTCVideoRotation_0
+                                                          timeStampNs: timeStampNs];
+
+    dispatch_async(dispatch_get_main_queue(), ^{
```
[the remark this answers is not in the document]
Comment (alexbumbu) [reply ios/RCTWebRTC/ScreenCapturer.m:197]: it's crashing when calling the delegate from a secondary thread

File: ios/RCTWebRTC/ScreenCapturer.m (hunk 2 above)
Comment (alexbumbu) [reply ios/RCTWebRTC/ScreenCapturer.m:191]: The WebRTC code is just reading the time stamp from the CMSampleBufferRef, which we don't have. It's either we compute it on the client side and send it as part of the message, or compute it here. I chose the latest.

Comment (alexbumbu) [reply ios/RCTWebRTC/ScreenCapturer.m]: Fixed. I replaced the implementation to use `mach_time` instead of `NSDate`

File: ios/RCTWebRTC/SocketConnection.m (hunk 4 above)
Comment (alexbumbu) [reply ios/RCTWebRTC/SocketConnection.m]: Fixed. Updated to run on the `clientSocket`

File: BroadcastPickerView.js (hunk 1 above)
Comment (alexbumbu) [reply BroadcastPickerView.js]: done. The file was renamed to ScreenCapturePickerView

File: ios/RCTWebRTC/SocketConnection.m (hunk 3 above)
Comment (alexbumbu) [reply ios/RCTWebRTC/SocketConnection.m]: fixed

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m (hunk 5 above)
Comment (alexbumbu) [reply ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m]: reverted

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m (hunk 7 above)
Comment (alexbumbu) [reply ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m]: reverted

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m (hunk 8 above)
Comment (alexbumbu) [reply ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m]: reverted

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m (hunk 9 above)
Comment (alexbumbu) [reply ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:285]: reverted

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m (hunk 6 above)
Comment (alexbumbu) [reply ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m]: reverted

File: ios/RCTWebRTC/SocketConnection.m [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    self.inputStream.delegate = nil;
+    self.outputStream.delegate = nil;
+        
+    [self.inputStream close];
+    [self.outputStream close];
+        
+    [self.networkThread cancel];
+    
+    close(self.serverSocket);
+    dispatch_source_cancel(self.listeningSource);
```
Comment (saghul) [inline ios/RCTWebRTC/SocketConnection.m:95]: I think we should do this before close.

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m [hunk 12]
```diff
@@ -45,13 +49,59 @@ - (RTCVideoTrack *)createVideoTrack:(NSDictionary *)constraints {
...
   return videoTrack;
 }
 
+- (RTCVideoTrack *)createScreenCaptureVideoTrack {
+    RTCVideoSource *videoSource = [self.peerConnectionFactory videoSource];
+
+    NSString *trackUUID = [[NSUUID UUID] UUIDString];
+    RTCVideoTrack *videoTrack = [self.peerConnectionFactory videoTrackWithSource:videoSource trackId:trackUUID];
+
+#if !TARGET_IPHONE_SIMULATOR
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:65]: Let's just return null here.

**Review by saghul — COMMENTED** [review]
Left some comments PTAL!

File: ios/RCTWebRTC/ScreenCapturePickerViewManager.m [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+//  react-native-webrtc
+//
+//  Created by Alex-Dan Bumbu on 20/01/2021.
+//
+
+#import <ReplayKit/ReplayKit.h>
+#import <React/RCTUIManager.h>
+
+#import "ScreenCapturePickerViewManager.h"
+#import "WebRTCModule.h"
```
[the remark this answers is not in the document]
Comment (alexbumbu) [reply ios/RCTWebRTC/ScreenCapturePickerViewManager.m:12]: removed

File: ios/RCTWebRTC/ScreenCapturePickerViewManager.m [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+@end
+
+@implementation ScreenCapturePickerViewManager {
+    RPSystemBroadcastPickerView *_broadcastPickerView;
+}
+
+RCT_EXPORT_MODULE()
+
+- (UIView *)view {
+    _broadcastPickerView = [[RPSystemBroadcastPickerView alloc] initWithFrame:kBroadcastPickerViewFrame];
```
[the remark this answers is not in the document]
Comment (alexbumbu) [reply ios/RCTWebRTC/ScreenCapturePickerViewManager.m:29]: done

File: ios/RCTWebRTC/ScreenCapturer.m [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+    
+    _readLength = [self.message appendBytes:buffer length:numberOfBytesRead];
+    if (_readLength == -1 || _readLength > kMaxReadLength) {
+        _readLength = kMaxReadLength;
+    }
+}
+
+- (void)didCaptureVideoFrame:(CVPixelBufferRef)pixelBuffer {
+    mach_timebase_info_data_t timebaseInfo;
+    mach_timebase_info(&timebaseInfo);
```
[the remark this answers is not in the document]
Comment (alexbumbu) [reply ios/RCTWebRTC/ScreenCapturer.m:188]: done

File: ios/RCTWebRTC/ScreenCapturer.m [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+    
+    if (_startTimeStampNs < 0) {
+        _startTimeStampNs = currentTimeStampNs;
+    }
+    
+    RTCCVPixelBuffer *rtcPixelBuffer = [[RTCCVPixelBuffer alloc] initWithPixelBuffer: pixelBuffer];
+    int64_t frameTimeStampNs = currentTimeStampNs - _startTimeStampNs;
+    
+    RTCVideoFrame *videoFrame = [[RTCVideoFrame alloc] initWithBuffer: rtcPixelBuffer
+                                                             rotation: RTCVideoRotation_0
```
[the remark this answers is not in the document]
Comment (alexbumbu) [reply ios/RCTWebRTC/ScreenCapturer.m:201]: done

File: ios/RCTWebRTC/SocketConnection.m [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+    [self setupNetworkThread];
+        
+    self.serverSocket = socket(AF_UNIX, SOCK_STREAM, 0);
+    if (self.serverSocket < 0) {
+       NSLog(@"failure creating socket");
+       return nil;
+    }
+    
+    if (![self setupSocketWithFileAtPath: filePath]) {
+        return nil;
```
[the remark this answers is not in the document]
Comment (alexbumbu) [reply ios/RCTWebRTC/SocketConnection.m:39]: fixed

File: ios/RCTWebRTC/SocketConnection.m (hunk 11 above)
Comment (alexbumbu) [reply ios/RCTWebRTC/SocketConnection.m:95]: done

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m (hunk 12 above)
Comment (alexbumbu) [reply ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:65]: done

File: ios/RCTWebRTC/ScreenCapturer.m [hunk 18]
```diff
@@ -0,0 +1,232 @@
...
+    
+    // Copy the pixel buffer
+    CVReturn status = CVPixelBufferCreate(kCFAllocatorDefault, width, height, kCVPixelFormatType_32BGRA, NULL, &_imageBuffer);
+    if (status != kCVReturnSuccess) {
+        NSLog(@"CVPixelBufferCreate failed");
+        return false;
+    }
+    
+    [self copyImageData:messageData toPixelBuffer:&_imageBuffer];
+    if (_imageBuffer == nil) {
```
[the remark this answers is not in the document]
Comment (alexbumbu) [reply ios/RCTWebRTC/ScreenCapturer.m]: removed

File: ios/RCTWebRTC/ScreenCapturePickerViewManager.m [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+//  react-native-webrtc
+//
+//  Created by Alex-Dan Bumbu on 20/01/2021.
+//
+
+#import <ReplayKit/ReplayKit.h>
+#import <React/RCTUIManager.h>
+
+#import "ScreenCapturePickerViewManager.h"
+//#import "WebRTCModule.h"
```
Comment (saghul) [inline ios/RCTWebRTC/ScreenCapturePickerViewManager.m:12]: ❓

**Review by saghul — APPROVED** [review]
Excellent excellent job here. LGTM. I'll give it a test before merge but it all checks out!
